### DISCUSSION TITLE
Model alias support

### DISCUSSION THREAD

**Discussion by dutchman1990** [spatie/laravel-permission#2428] [Q&A] [answered]
Do we have any way to use Model alias in custom guard.

```
'providers'        => [
        'users'           => [
            'driver' => 'eloquent',
            'model'  => App\User::class
        ],
        'subdealers' => [
            'driver' => 'eloquent',
            'model'  => \SubDealers::class
        ]
    ]
```
This is my auth.php. Where **\SubDealers::class** is an alias of **tbl_subdealer_master** model which is available in 
```
App\Models\Partners\Subdealer\tbl_subdealer_master
```

If I use the alias as model_type in roles table laravel does not attach role object in auth, but if I use the full model path it works. I would like to use the alias. Am I missing something?

**Comment by drbyte** [comment]
I wonder if it would work if you were to resolve the alias inside the `getModelForGuard()` helper function in this package's `/src/helpers.php` file?

**Comment by drbyte** [comment]
> Where **\SubDealers::class** is an alias of **tbl_subdealer_master** model which is available in
> `App\Models\Partners\Subdealer\tbl_subdealer_master`

Are you registering it as a facade? or as an app alias? or how?
Can you show your application code where you define/register the alias, and the model with its headers and constructor?

**Reply by dutchman1990** [reply]
Subdealers::class is written like below in app.php

```
'aliases'            => Facade::defaultAliases()->merge([ ...
      'SubDealers' => App\Models\Partners\Subdealer\tbl_subdealer_master::class
...
])->toArray()
```

here is the model

```
class tbl_subdealer_master extends Authenticatable {
    use SoftDeletes, HasApiTokens, HasRoles;
    protected $primaryKey = 'subdealer_id';
    protected $guarded    = ['subdealer_id'];
    protected $casts      = [
        'products' => 'array'
    ];
}
````

Usually I use alias name in morph relation based tables so that I can change the model path without breaking the existing code. Maybe I am doing it wrong for this case.

**Reply by drbyte** [reply]
> Usually I use alias name in morph relation based tables so that I can change the model path without breaking the existing code.

I've seen people use `enforceMorphMap` for that.
https://laravel.com/docs/10.x/eloquent-relationships#custom-polymorphic-types

**Reply by drbyte** [reply]
I'm trying to come up with a suitable test to go with a possible solution so that you don't need to write a custom morphMap.

I was hoping that just calling `getAlias()` would suffice, but I think that's the opposite of what you're after (eg: it would store the resolve actual FQDN instead of the alias).

**Reply by drbyte** [reply]
Apologies. I believe I understand your goal, but admittedly I've had no cause for using such aliases in an extensive way, and therefore only cursory understanding of the technical implementation required behind-the-scenes for this. After spending many hours exploring, and repeatedly feeling like I was "almost there", I've not come up with something that could be built-in to the package. At least, not without another set of eyes and technical expertise on the subject.

I'm inclined to suggest a manual implementation of `enforceMorphMap` ;)

**Reply by drbyte** [reply]
Does specifying `getMorphClass` within the model work for your requirements?
Ref: https://spatie.be/docs/laravel-permission/v5/advanced-usage/extending#content-child-user-models

**Reply by dutchman1990** [reply, marked as answer]
Sorry for the late response. Found another way to bind that with model. Which is written as secondary solution.

**Solution 1** 
With **enforceMorphMap**
I had to write below code in AppServiceProvider.php 's boot method.

```
use Illuminate\Database\Eloquent\Relations\Relation; 

public function boot() {
        Relation::enforceMorphMap([
            'SubDealers' => 'App\Models\Partners\Subdealer\tbl_subdealer_master'
        ]);
    }
```
and below line in model file.

```
protected $guard_name = 'web'; //this could be a custom guard. I choose to go with default one.
```
**Solution 2 - Secondary Solution** 
I had to append one model attribute name as id to bind roles object with the model. This one may have flaws. I Would recommend solution 1 though. 

```
protected $appends = ['id'];

public function getIdAttribute($value) {
      return $value->subdealer_id;
}
```
